Core: Add logLevel preset property to filter logging

ndelangen opened the pull request:
Issue: #8461

## What I did

ADD logLevel preset property (default is `info`)
ADD global variable injection
ADD filter for lower level log messages
ADD a global method for changing the logLevel at runtime.

ndelangen committed:
ADD logLever preset property && filter lower lever log messages

ndelangen committed:
FIX lockfile

ndelangen committed:
FIX tests

shilman commented on addons/links/src/preview.test.js:
```diff
@@ -17,6 +17,8 @@ jest.mock('global', () => ({
     })),
   },
+  window: global,
+  __STORYBOOK_LOGGER: 'warn',
```
`__STORYBOOK_LOG_LEVEL__`?

shilman replied:
what is this supposed to do? below `__STORYBOOK_LOGGER` is set to `console`, which is a different type?

shilman commented on lib/core/src/server/preview/iframe-webpack.config.js:
```diff
@@ -39,6 +39,7 @@ export default async ({
   presets,
 }) => {
   const dlls = await presets.apply('webpackDlls', []);
+  const logLevel = await presets.apply('logLevel', []);
```
Is this so the user can set `logLevel` in `main.js`? what happens when it gets defined in a preset? Is there one level globally, or does it get scoped somehow?

ndelangen replied:
The value is a singular string, not an array.

There's 1 setting, whichever is last wins, `main.js` should always be last.

presets should not be setting this I think.

shilman reviewed (commented):
LGTM but a few questions

ndelangen committed:
Merge branch 'next' into add/loglevel

ndelangen committed:
Merge branch 'next' into add/loglevel

ndelangen committed:
FIX

ndelangen committed:
Set correct initial value of logLevel

ndelangen committed:
Merge branch 'tech/install-optional-on-ci' into add/loglevel

ndelangen committed:
ADD puppeteer as optionalDependency

ndelangen committed:
FIX

ndelangen committed:
FIX

ndelangen committed:
Merge branch 'next' into add/loglevel

ndelangen commented:
No idea why puppeteer is all of a sudden not able to find the globals anymore, and use them to extract.

When I run it locally it works.

ndelangen committed:
Merge branch 'next' into add/loglevel

# Conflicts:
#	yarn.lock

ndelangen committed:
Merge branch 'next' into add/loglevel

ndelangen committed:
Merge branch 'next' into add/loglevel

# Conflicts:
#	yarn.lock

ndelangen committed:
ADD a loglevel option for dev mode in CLI

it's not used anywhere I think, but we're getting quite a few 'invalid cli option' errors,
and this should get rid of them.

ndelangen committed:
Merge branch 'next' into tech/loglevel-for-dev

ndelangen committed:
Merge branch 'next' into tech/loglevel-for-dev

tmeasday commented on lib/client-logger/src/index.ts:
```diff
@@ -1,16 +1,25 @@
+  trace: (message: any, ...rest: any[]): void => console.trace(message, ...rest),
   debug: (message: any, ...rest: any[]): void => console.debug(message, ...rest),
-  log: (message: any, ...rest: any[]): void => console.log(message, ...rest),
+  log: (message: any, ...rest: any[]): void => window.console.log(message, ...rest),
```
This is pretty surprising, what's up with this?

ndelangen replied:
`log` has no "level" to it. it's not part of `console` lib.

I could decide to make `log` equal to `info`?

tmeasday commented on lib/client-logger/src/index.ts:
```diff
@@ -1,16 +1,25 @@
+}
+
+// eslint-disable-next-line no-underscore-dangle
+window.__STORYBOOK_LOGGER = console;
```
Should we just check if `window` is defined before setting this?

ndelangen replied:
✅

tmeasday commented:
@ndelangen I think the most likely situation is that the storybook is throwing an error as it loads in puppeteer and we aren't seeing the error. Try adding:

```js
  this.page.on('console', (msg) => {
    const type = msg.type();
    logger.info({ browserConsole: true, type, ...msg.location() }, msg.text());
  });
```

Either that or get a build storybook off one of the CI machines (I have one here)

ndelangen committed:
Merge branch 'next' into add/loglevel

# Conflicts:
#	lib/cli/src/extract.ts
#	lib/client-logger/src/index.ts
#	yarn.lock

ndelangen committed:
FIX loglevel being a string within a string

ndelangen committed:
FIX lockfile && safety check for window

shilman reviewed (approved):
LGTM does this need documentation somewhere?

ndelangen committed:
Merge branch 'next' into add/loglevel

ndelangen committed:
ADD docs

ndelangen committed:
Merge branch 'tech/loglevel-for-dev' into add/loglevel

ndelangen committed:
set loglevel for node based on cli param in dev-mode

ndelangen committed:
Use cli param to set default loglevel in preset

ndelangen commented:
I added documentation and the ability to set a value for this from the cli
